File: src/Core/DownloadWorkflow.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using ICSharpCode.SharpZipLib.GZip;
using ICSharpCode.SharpZipLib.Tar;
using ICSharpCode.SharpZipLib.Zip;
using NPackage.Core.Extensions;

namespace NPackage.Core
{
    public class DownloadWorkflow
    {
        private class DownloadAction
        {
            private readonly Uri uri;
            private readonly string filename;
            private readonly Action<string> continuation;

            public DownloadAction(Uri uri, string filename, Action<string> continuation)
            {
                this.uri = uri;
                this.filename = filename;
                this.continuation = continuation;
            }

            public Uri Uri
            {
                get { return uri; }
            }

            public string Filename
            {
                get { return filename; }
            }

            public Action<string> Continuation
            {
                get { return continuation; }
            }
        }

        private readonly List<DownloadAction> actions = new List<DownloadAction>();

        public void Enqueue(Uri uri, string filename, Action<string> continuation)
        {
            actions.Add(new DownloadAction(uri, filename, continuation));
        }

        private static string AddFilename(string pathOrFilename, string filename)
        {
            bool isPath = pathOrFilename.EndsWith(Path.DirectorySeparatorChar.ToString())
                || pathOrFilename.EndsWith(Path.AltDirectorySeparatorChar.ToString());

            return isPath
                ? Path.Combine(pathOrFilename, filename)
                : pathOrFilename;
        }

        public bool Step()
        {
            var actionsByFilenameByUri = actions
                .GroupBy(a => a.Uri)
                .Select(g => new
                                 {
                                     Uri = g.Key,
                                     ActionsByFilename = g
                                 .GroupBy(a => a.Filename, StringComparer.InvariantCultureIgnoreCase)
                                 .Select(g2 => new { Filename = g2.Key, Actions = g2.ToArray() })
                                 .OrderBy(a => a.Filename, StringComparer.InvariantCultureIgnoreCase)
                                 .ToArray()
                                 })
                .OrderBy(a => a.Uri.ToString())
                .ToArray();

            actions.Clear();

            foreach (var uriPair in actionsByFilenameByUri)
            {
                string firstFilename = null;
                FileInfo firstFileInfo = null;
                string responseUriFilename = null;

                foreach (var filenamePair in uriPair.ActionsByFilename)
                {
                    if (firstFilename == null)
                    {
                        string responseFilename;
                        RaiseLog("Checking {0}", uriPair.Uri);

                        WebRequest request = WebRequest.Create(uriPair.Uri);
                        using (WebResponse response = request.GetResponse())
                        {
                            responseUriFilename = Path.GetFileName(response.ResponseUri.GetComponents(UriComponents.Path, UriFormat.Unescaped));
                          
                            string contentDisposition = response.Headers["Content-Disposition"];
                            if (!String.IsNullOrEmpty(contentDisposition))
                            {
                                string[] parts = contentDisposition.Split(new[] { ';' }, 2);
                                if (parts.Length > 1)
                                {
                                    string part1 = parts[1].TrimStart();
                                    const string prefix = "filename=";
                                    if (part1.StartsWith(prefix))
                                        responseUriFilename = part1.Substring(prefix.Length).Trim('"').Trim();
                                }
                            }

                            responseFilename = AddFilename(filenamePair.Filename, responseUriFilename);
                            FileInfo fileInfo = new FileInfo(responseFilename);

                            HttpWebResponse httpWebResponse = response as HttpWebResponse;
                            if (!fileInfo.Exists || (httpWebResponse != null && httpWebResponse.LastModified > fileInfo.LastWriteTime))
                            {
                                using (Stream inputStream = response.GetResponseStream())
                                {
                                    RaiseLog("Downloading from {0} to {1}", response.ResponseUri, responseFilename);
                                    inputStream.CopyTo(responseFilename);
                                    fileInfo.Refresh();

                                    if (httpWebResponse != null)
                                        fileInfo.LastWriteTime = httpWebResponse.LastModified;
                                }
                            }
                            
                            firstFilename = responseFilename;
                            firstFileInfo = fileInfo;
                        }
                        
                        foreach (DownloadAction action in filenamePair.Actions)
                            action.Continuation(responseFilename);
                    }
                    else
                    {
                        string responseFilename = AddFilename(filenamePair.Filename, responseUriFilename);
                        FileInfo fileInfo = new FileInfo(responseFilename);

                        if (!fileInfo.Exists || firstFileInfo.LastWriteTime > fileInfo.LastWriteTime)
                        {
                            RaiseLog("Copying from {0} to {1}", firstFilename, responseFilename);
                            File.Copy(firstFilename, responseFilename, true);
                            fileInfo.LastWriteTime = firstFileInfo.LastWriteTime;
                        }

                        foreach (DownloadAction action in filenamePair.Actions)
                            action.Continuation(responseFilename);
                    }
                }
            }
            
            return actions.Count > 0;
        }

        private void RaiseLog(string format, params object[] args)
        {
            EventHandler<LogEventArgs> handler = Log;
            if (handler != null)
                handler(this, new LogEventArgs(String.Format(format, args)));
        }

        private static InvalidOperationException NotFoundInArchive(string archiveFilename, string entryName)
        {
            string message = String.Format("There is no {0} in {1}.", entryName, archiveFilename);
            throw new InvalidOperationException(message);
        }

        public static void ExtractFile(string archiveFilename, string entryName, string filename)
        {
            if (archiveFilename.EndsWith(".zip", StringComparison.InvariantCultureIgnoreCase))
            {
                using (ZipFile file = new ZipFile(archiveFilename))
                {
                    int index = file.FindEntry(entryName, true);
                    if (index < 0)
                        throw NotFoundInArchive(archiveFilename, entryName);

                    using (Stream stream = file.GetInputStream(index))
                        stream.CopyTo(filename);
                }
            }
            else if (archiveFilename.EndsWith(".tar.gz", StringComparison.InvariantCultureIgnoreCase) ||
                     archiveFilename.EndsWith(".tgz", StringComparison.InvariantCultureIgnoreCase))
            {
                using (Stream fileStream = File.Open(archiveFilename, FileMode.Open, FileAccess.Read))
                using (Stream gzipStream = new GZipInputStream(fileStream))
                using (TarInputStream tarStream = new TarInputStream(gzipStream))
                {
                    TarEntry entry;
                    while ((entry = tarStream.GetNextEntry()) != null)
                    {
                        if (String.Equals(entry.Name, entryName, StringComparison.InvariantCultureIgnoreCase))
                        {
                            using (FileStream outputStream = File.Create(filename))
                                tarStream.CopyEntryContents(outputStream);

                            break;
                        }
                    }

                    if (entry == null)
                        throw NotFoundInArchive(archiveFilename, entryName);
                }
            }
            else
                throw new NotSupportedException(archiveFilename + " is not a recognised archive.");
        }

        public event EventHandler<LogEventArgs> Log;
    }
}

File: src/Core/DownloadWorkflow.fs
namespace NPackage.Core

open System
open System.Collections.Generic
open System.IO
open System.Linq
open System.Net
open System.Text

type DownloadAction = { Key : UriKey;
                        Continuation : string -> unit }

type DownloadWorkflow(log) =
    let actions = new ResizeArray<DownloadAction>()

    let sanitise (s : String) =
        let invalidPathChars = 
            Path.InvalidPathChars
            |> Seq.append [| Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar, ':' |]
            |> Array.ofSeq

        let rec sanitise' (sb : StringBuilder) =
            match sb.ToString().IndexOfAny(invalidPathChars) with
            | n when n < 0 -> sb
            | n ->
                sb.[n] <- '-'
                sanitise' sb

        (sanitise' (new StringBuilder(s))).ToString()

    let ensureFilename (pathOrFilename : string) (uri : Uri) filename =
        match pathOrFilename.EndsWith(Path.DirectorySeparatorChar.ToString()) || pathOrFilename.EndsWith(Path.AltDirectorySeparatorChar.ToString()) with
        | true -> 
            let dir1 =
                let authority = uri.GetLeftPart(UriPartial.Authority)
                if authority = String.Format("{0}://{1}", Uri.UriSchemeHttp, uri.Host) then
                    uri.Host
                else
                    authority

            let dir2 = Path.GetDirectoryName(uri.PathAndQuery.TrimStart('/'))
            Path.Combine(pathOrFilename, Path.Combine(sanitise dir1, Path.Combine(sanitise dir2, filename)))
        | false -> pathOrFilename

    let attachmentFilename (response : WebResponse) =
        let contentDisposition = response.Headers.["Content-Disposition"]
        if String.IsNullOrEmpty(contentDisposition) then
            None
        else
            let parts = contentDisposition.Split([| ';' |])
            if parts.Length > 1 then
                let part1 = parts.[1].TrimStart()
                let prefix = "filename="
                if part1.StartsWith(prefix) then
                    Some (part1.Substring(prefix.Length).Trim('"').Trim())
                else
                    None
            else
                None

    let downloadFirst (uri : Uri) filename actions =
        log (String.Format("Checking {0}", uri))

        let request = WebRequest.Create(uri)
        match request with
        | :? HttpWebRequest as httpWebRequest -> 
            try
                httpWebRequest.UseDefaultCredentials <- true
            with :? NotImplementedException ->
                ()

        | _ -> ()

        use response = request.GetResponse()
        let responseUriFilename =
            match attachmentFilename response with
            | Some s -> s
            | None -> Path.GetFileName(response.ResponseUri.GetComponents(UriComponents.Path, UriFormat.Unescaped))
                          
        let responseFilename = ensureFilename filename uri responseUriFilename
        let fileInfo = new FileInfo(responseFilename)

        let lastModified =
            match response with
            | :? HttpWebResponse as httpWebResponse -> httpWebResponse.LastModified
            | :? FileWebResponse as fileWebResponse -> File.GetLastWriteTimeUtc(fileWebResponse.ResponseUri.LocalPath)
            | _ -> DateTime.UtcNow

        if (not fileInfo.Exists) || (lastModified > fileInfo.LastWriteTimeUtc) then
            log (String.Format("Downloading from {0} to {1}", response.ResponseUri, responseFilename))
            using (response.GetResponseStream()) (fun inputStream ->
                let dir = Path.GetDirectoryName(responseFilename)
                ignore (Directory.CreateDirectory(dir))
                use outputStream = File.Create(responseFilename)
                StreamExtensions.copy inputStream outputStream)

            fileInfo.Refresh()
            fileInfo.LastWriteTimeUtc <- lastModified

        for action in actions do
            action.Continuation responseFilename

        (responseFilename, fileInfo, responseUriFilename)

    member this.Enqueue(key, continuation) =
        actions.Add({ Key = key; Continuation = continuation })

    member this.Step() =
        let actionsByFilenameByUri = 
            actions
            |> Seq.groupBy (fun { Key = { Uri = uri } } -> uri)
            |> Seq.map (fun (key, g) -> 
                (key, g
                        |> Seq.groupBy (fun { Key = { Filename = filename } } -> filename)
                        |> Seq.map (fun (key', g') -> (key', List.ofSeq g'))
                        |> List.ofSeq
                        |> List.sortBy fst))
            |> List.ofSeq
            |> List.sortBy (fun (key, _) -> key.ToString())

        actions.Clear()

        for (uri, actionsByFilename) in actionsByFilenameByUri do
            match actionsByFilename with
            | [] -> ()
            | [(filename, actions)] -> ignore (downloadFirst uri filename actions)
            | (filename, actions) :: rest ->
                let (firstFilename, firstFileInfo, responseUriFilename) = downloadFirst uri filename actions
                let firstWriteTime = firstFileInfo.LastWriteTimeUtc
                for (filename, actions) in rest do
                    let responseFilename = ensureFilename filename uri responseUriFilename
                    let fileInfo = new FileInfo(responseFilename)

                    if (not fileInfo.Exists) || firstWriteTime > fileInfo.LastWriteTimeUtc then
                        log (String.Format("Copying from {0} to {1}", firstFilename, responseFilename))
                        let dir = Path.GetDirectoryName(responseFilename)
                        ignore (Directory.CreateDirectory(dir))
                        File.Copy(firstFilename, responseFilename, true)
                        fileInfo.LastWriteTimeUtc <- firstWriteTime

                    for action in actions do
                        action.Continuation responseFilename
            
        actions.Count > 0